pkg/controllers/user/approuter/core_dns_client.go

```diff
@@ -0,0 +1,290 @@
+package approuter
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"sort"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/rdns-server/model"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/sirupsen/logrus"
+	k8scorev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	contentType     = "Content-Type"
+	jsonContentType = "application/json"
+	secretKey       = "rdns-token"
+)
+
+func jsonBody(payload interface{}) (io.Reader, error) {
+	buf := &bytes.Buffer{}
+	err := json.NewEncoder(buf).Encode(payload)
+	if err != nil {
+		return nil, err
+	}
+	return buf, nil
+}
+
+type Client struct {
+	httpClient             *http.Client
+	base                   string
+	lock                   *sync.RWMutex
+	managementSecretLister v1.SecretLister
+	secrets                v1.SecretInterface
+	clusterName            string
+}
+
+func (c *Client) request(method string, url string, body io.Reader) (*http.Request, error) {
+	req, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add(contentType, jsonContentType)
+
+	return req, nil
+}
+
+func (c *Client) do(req *http.Request) (model.Response, error) {
+	var data model.Response
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return data, err
+	}
+	// when err is nil, resp contains a non-nil resp.Body which must be closed
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return data, errors.Wrap(err, "Read response body error")
+	}
+
+	err = json.Unmarshal(body, &data)
+	if err != nil {
+		return data, errors.Wrap(err, "Decode response error")
+	}
+	logrus.Debugf("Got response entry: %+v", data)
+	if code := resp.StatusCode; code < 200 || code > 300 {
```

Comment:
looks like checking response code before reading the body can save some on performance. Can we move it up?

Reply:
Yep,I think what you say is very reasonable.However,I think about that we want to get error message，still need to read the body first.
```
if code := resp.StatusCode; code < 200 || code > 300 {
		if data.Message != "" {
			return data, errors.Errorf("Got request error: %s", data.Message)
		}
	}
```